fix(vmgen): wrong `range` value being read

zerbina opened the pull request:
## Summary

Fix an incorrect/invalid value being read from locations of unsigned
`range` type. Only the VM backend was affected, and only when the value
stored in the location had the most-significant bit set.

## Details

With the VM, all unsigned values require masking after being read from
a memory location.  `genRegLoad`  emits the necessary masking
instruction,
but `isUnsigned` doesn't consider range types (a separate bug), meaning
that values of type `tyRange` with an unsigned base type were not
masked.

* all code generation for narrowing/masking instructions is changed to
  skip `tyRange` types
* the custom narrowing implementation used for `mShrI` code generation
  is merged into `genNarrow`
* the duplicated unsigned integer masking is moved to a dedicated
  procedure (`genNarrowUnsigned`)

There was also the problem of the types' `size` field being accessed
directly; since a type's size is computed on demand, the value in
`size` might not be valid. `getSize` is used instead.

zerbina committed:
fix(vmgen): wrong range value being read

Summary
=======

Fix an incorrect/invalid value being read from locations of unsigned
`range` type. Only the VM backend was affected, and only when the value
stored in the location had the most-significant bit set.

Details
=======

With the VM, all unsigned values require masking when being read from a
memory location. `genRegLoad` emits the necessary masking instruction,
but `isUnsigned` doesn't consider range types (a separate bug), meaning
that `tyRange` with an unsigned base type were not masked.

* all code generation for narrowing/masking instruction is changed to
  skip `tyRange` types
* the custom narrowing implementation used for `mShrI` code generation
  is merged into `genNarrow`
* the duplicated unsigned integer masking in is moved to a dedicated
  procedure (`genNarrowUnsigned`)

There was also the problem of the type's `size` field being accessed
directly; since type size are populated on demand, the value in `size`
might not be valid. `getSize` is used instead.

saem commented on compiler/vm/vmgen.nim:
```diff
@@ -1337,22 +1337,33 @@ proc genBinaryABC(c: var TCtx; n: CgNode; dest: var TDest; opc: TOpcode) =
-    c.gABC(n, opcNarrowU, dest, TRegister(t.size*8))
-  elif t.kind in {tyInt8..tyInt32} or (t.kind == tyInt and t.size < 8):
-    c.gABC(n, opcNarrowS, dest, TRegister(t.size*8))
+proc genNarrow(c: var TCtx; n: CgNode; dest: TRegister; sNarrow = opcNarrowS) =
```
```suggestion
proc genNarrow(c: var TCtx; info: TLineInfo; dest: TRegister; sNarrow = opcNarrowS) =
```

This could be a line info, instead of `n: CgNode`, in which case `genNarrowU` and `genNarrowUnsigned` could be simplified out or become inline `proc`s which specify the opcode and possibly carry some minimal logic.

zerbina replied:
Only `genNarrowU` can be simplified away this way; `genNarrowUnsigned` deviates from the other narrow procedures in that it generates nothing for signed types.

I'll replace `genNarrowU`.

saem replied:
I'm not sure how I feel about it, as it could be too clever, but `ocpRet` could potentially be used as a sentinel value to indicate "just return the value if signed"?

It's probably not a great idea.

zerbina replied:
> but `ocpRet` could potentially be used as a sentinel value to indicate "just return the value if signed"?

Yep, that would work, although I think I'd rather leave `genNarrowUnsigned` be for the moment.

saem reviewed (approved):
One soft stylistic suggestion that might reduce some code, otherwise looks good.

zerbina committed:
vmgen: fix the size of the wrong `PType` being queried

This had no effect on program behaviour, but it could result in
unnecessary type size computations.

zerbina committed:
vmgen: dispatch to `genNarrow` from `genNarrowU`

saem commented:
/merge
